### DISCUSSION TITLE
Nuxt3 dynamic <component>

### DISCUSSION THREAD

**Discussion by Bartas139** [nuxt/nuxt#26763] [Questions]
Hi, can someone help me with creating a dynamic component based on JSON data in NUXT3.

I have DataTable body component like this:

```
<tbody class="divide-y divide-gray-200">
          <tr v-for="item in data" :key="item._id">
            <td v-for="column in columns" class="whitespace-nowrap px-3 py-4 text-sm text-gray-500">
              <div v-if="column.definition">
                <component :is="resolveComponent(column?.definition(item[column.key]).component)" v-bind="column?.definition(item[column.key]).component.props" >
                  {{ column?.definition(item[column.key]).slot }}
                </component>
              </div>
              <div v-else>{{ item[column.key] }}</div>
            </td>
          </tr>
        </tbody>
```
As you can see I want to render a component, where the` :is `prop is defined in `column?.definition(item[column.key]).component`

The definition function return an object and the .component property contains a simple string, as you can see here `{title: 'Email', key: 'email', definition: (item)=>{return {component: 'GeneralButton', props: {to: item._id}, slot: item}}}
`
So the result in this exmaple, which should be rendered is actualy:

`<component :is="resolveComponent('GeneralButton')" ... >`
And thats the problem, the resolveComponent doesnt work while I pass a variable in it. In case I pass there directly the same string, which is returned by definition function it works great.

I tried some other options, but none of them worked. The most used one is to use a global component, so the GeneralButton should be global. This aproach works, but I want to create a fully customisable component and I actualy dont know, which component will I want to use in future so this is not a good solution for me. (I saw also an advice to set all components as global and I want to avoid this cause of performance issues).

Other aproach was to override Nuxt3 auto import and manualy import desired component to DataTableBody component. This alse works and the resolveComponent() function than works with varibale :is, but it has the same problem as the previous one, I dont know which component I will get in definition() function. This should Actualy work, if there was a way to define the import dynamicly based on variable, but I am quite sure it is no possible, something like:

`import {`$componentToRender`} from '#components' //This does not work and I didnt find any solution which works`

Is there actualy a way to get this work? Thanks :)

**Comment by AloisSeckar** [comment]
It is [pointed out in the docs](https://nuxt.com/docs/guide/directory-structure/components#dynamic-components), that you only can pass in a literal string into Nuxt's `resolveComponent` and not any variable

The workaround known to me is to make a custom wrapper to translate dynamic variable into concrete strings:

```
function getDynamicComponent(componentName: string) {
  switch (componentName) {
    case 'component1':
      return resolveComponent('component1');
    case 'component2':
      return resolveComponent('component2');
    // ...
  }
}

The downside is you have to know the finite list of possible values in advance.

**Comment by Bartas139** [comment]
> It is [pointed out in the docs](https://nuxt.com/docs/guide/directory-structure/components#dynamic-components), that you only can pass in a literal string into Nuxt's `resolveComponent` and not any variable
> 
> The workaround known to me is to make a custom wrapper to translate dynamic variable into concrete strings:
> 
> ```
> function getDynamicComponent(componentName: string) {
>   switch (componentName) {
>     case 'component1':
>       return resolveComponent('component1');
>     case 'component2':
>       return resolveComponent('component2');
>     // ...
>   }
> }
> 
> The downside is you have to know the finite list of possible values in advance.
> ```

Hi, thanks for quick response, thats actualy the problem I am heading - I Don't know the possible values. Is there any other workaround? Or maybe any other way to aproach my goal, I actualy don't need to use resolveComponent, I would be happy with any other possible way to 'dynamicly call a component' specified in variable.

**Reply by AloisSeckar** [reply]
Looks like some approaches are mentioned in #14036

**Reply by Bartas139** [reply]
@AloisSeckar  Yes I saw that discussion and many others, they are all suggesting to use:

- Nuxt3 buildin way to register global components (./components/global)
- Register specific component as global in nuxt config
**_Both of these require to know possible components_**
 
OR

- Register entire components folder as global in nuxt config
- Use `import * as components from '#components'`
**_I am rly aware of performance issues with this aproach_**

It actualy seems this has no clear solution in my eyes.
